### PR TITLE
PDF parsing, EPUB Experiments

### PR DESCRIPTION
**Pull request by ManchurianMan**
## **tl;dr** _we're going to have to do our work with epubs_, not pdfs or html files. PDF will be necessary for page numbers only.

After many attempts to figure out how to make PDFs work, I just kept running into too many problems. I checked the website's HTML files, but they were basically useless. They didn't have crucial things like calendars and table of contents, which doesn't solve the main problem I wanted to solve with HTML files, namely not having to use PDFs.

Turns out there's a python library for epubs and there was an epub on the website. It contains all the info we need to make this work except page numbers. Epubs can be a pain to work with, though.

### **Next steps: what are we even making?**

We need to decide what we want the final JSON to look like. If we just want the raw text, we have the epub, so we can stop now. How do we want it augmented? If we think about this as creating an API that takes an input of a day and returns certain texts, we need to define how those texts are returned before we start coding. This is an exercise in [design-first API creation](https://devblogs.microsoft.com/ise/2023/05/08/design-api-first-with-typespec/).

### CODE REVIEW

**Comment by ax-wwidner** [comment]
I like the idea of treating this as an exercise in API design. And yes, most of it is date-based. I think we can break it down even further, though. 

1. An endpoint that returns the current Sunday we're in. Ie. the 18th Sunday after trinity. Bonus if this endpoint optionally accepted a date, assuming today's date if no date is given, and gave you back the Sunday for the given date. 
2. An endpoint that returns the Psalm reading of the day. Verse and also the page number in the 1928 prayer book.
3. Endpoints returning the collect, NT, and OT readings.
4. I'm not sure how we'll handle this but we need an endpoint that returns the "extra" bits. Perhaps this endpoint can accept an input of morning, evening, family morning, family evening, shorter, compline and return the blocks of text in the order they should be read

**Comment by kai5263499** [comment]
I like the idea of treating this as an exercise in API design. And yes, most of it is date-based. I think we can break it down even further, though.

1. An endpoint that returns the current Sunday we're in. Ie. the 18th Sunday after trinity. Bonus if this endpoint optionally accepted a date, assuming today's date if no date is given, and gave you back the Sunday for the given date.
2. An endpoint that returns the Psalm reading of the day. Verse and also the page number in the 1928 prayer book.
3. Endpoints returning the collect, NT, and OT readings.
4. I'm not sure how we'll handle this but we need an endpoint that returns the "extra" bits. Perhaps this endpoint can accept an input of morning, evening, family morning, family evening, shorter, compline and return the blocks of text in the order they should be read

**Comment by ManchurianMan** [comment]
Here are my thoughts as far as steps go, to keep this reasonable for newbie programmers like me:

1. Normal digitization. Basically just repackage the epub as a JSON.
2. Linked digitization. Make it so that we have a working table of contents for the JSON.
3. API backend. Give us a date, and we'll give you the prayer; give us an event, and we'll give you the associated text. This will be where most of the branching logic comes.*

Most of my thoughts have been cluttered because I've been lumping these together in my head. Step 1 is hard enough, because it involves appropriately treating the italicized text as instructions and knowing where to put it in the file structure.

_*This is where your comment comes in, Wes. I think once we have the linked digitization, the construction of the date logic starts, and the steps and order you gave make sense to me._

**Comment by ManchurianMan** [comment]
It's time to review if we're on the right track here. Most of the xhtml files (the meat of an epub) are converted into JSON, as can be seen. What remains is 1) going through and catching all the exceptions (there are a lot of them, it is the prayer book after all), and 2) adding in the other years for each Sunday (that's why there are multiple readings for a single day sometimes). 

2 should be relatively straightforward; 1, a pain in the neck. I'll likely concoct some test to see how much text from the original xhtml file I'm missing and "score" each one of the [JSON files](https://github.com/ctk-app-dev/simeon-gpt/tree/ben-pdf-parser/assets/db/json) so we have a nice metric for how close we are to being done. But before I go on fine-tuning, we have to decide whether this is the API structure we want. Here's an example:
```
{
  "Trinity Sunday": {
    "Sunday": {
      "Morning": [
        "Ps. 29, 99",
        "*Isa. 6:1-8",
        "1 Pet. 1:1-12"
      ],
      "Evening": [
        "Ps. 98, 100",
        "Ecclus. 43:1-12,27-33",
        "Eph. 4:1-16"
      ]
    },

```
Is that in the right ballpark? As long as the date->liturgical calendar turns out something like "wednesday after trinity sunday," or better yet "{week: trinity sunday, day: wednesday}",it'll play nicely with this structure. Otherwise I want to do the restructuring before the fine-tuning.

**Review by kai5263499 — APPROVED** [review]
This is a good start. Go ahead and merge this so we can close this branch. One of the more immediate next steps would be to focus on
1. The quality of the parsed data. It looks like you've got a lot in here accounting for data oddities you've encountered. that makes me think there is more polishing that's left to do
2. The structure of the data. I like that you are parsing this to json files. There is a lot we can do with those. What I'd like to think more about is file layout and also the shape of the json document that's produced and how well that works in applications like a BCP app and a liturgy-planning app.
